File: olorin-server/app/service/template_service.py
"""
Template Service
Feature: 005-polling-and-persistence

Provides CRUD operations and template application for investigation templates.
Supports template management, usage tracking, and placeholder replacement.

SYSTEM MANDATE Compliance:
- No hardcoded values: All configuration from environment
- Complete implementation: No placeholders or TODOs
- Type-safe: All parameters and returns properly typed
"""

from sqlalchemy.orm import Session
from typing import List, Dict, Any, Optional
from fastapi import HTTPException
from datetime import datetime
import json

from app.models.investigation_template import InvestigationTemplate
from app.models.investigation_state import InvestigationState
from app.schemas.investigation_state import InvestigationSettings
from app.service.template_helper import replace_placeholders, apply_overrides


class TemplateService:
    """Service for template CRUD operations and application."""

    def __init__(self, db: Session):
        self.db = db

    def list_templates(
        self,
        user_id: str,
        tags: Optional[List[str]] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> List[Dict[str, Any]]:
        """List user templates with filtering and pagination."""
        query = self.db.query(InvestigationTemplate).filter(
            InvestigationTemplate.user_id == user_id
        )

        if tags:
            for tag in tags:
                query = query.filter(InvestigationTemplate.tags.contains([tag]))

        templates = (
            query.order_by(InvestigationTemplate.last_used.desc())
            .limit(limit)
            .offset(offset)
            .all()
        )

        return [template.to_dict() for template in templates]

    def create_template(
        self,
        user_id: str,
        name: str,
        description: str,
        template_json: Dict[str, Any],
        tags: List[str],
    ) -> Dict[str, Any]:
        """Create template with settings validation."""
        InvestigationSettings(**template_json)

        template = InvestigationTemplate(
            user_id=user_id,
            name=name,
            description=description,
            template_json=json.dumps(template_json),
            tags=json.dumps(tags),
            usage_count=0,
            version=1,
        )

        self.db.add(template)
        self.db.commit()
        self.db.refresh(template)

        return template.to_dict()

    def get_template(self, template_id: str, user_id: str) -> Dict[str, Any]:
        """Get template by ID."""
        template = (
            self.db.query(InvestigationTemplate)
            .filter(
                InvestigationTemplate.template_id == template_id,
                InvestigationTemplate.user_id == user_id,
            )
            .first()
        )

        if not template:
            raise HTTPException(
                status_code=404, detail=f"Template {template_id} not found"
            )

        return template.to_dict()

    def update_template(
        self,
        template_id: str,
        user_id: str,
        name: Optional[str] = None,
        description: Optional[str] = None,
        template_json: Optional[Dict[str, Any]] = None,
        tags: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        """Update template. Increments version if JSON changed."""
        template = (
            self.db.query(InvestigationTemplate)
            .filter(
                InvestigationTemplate.template_id == template_id,
                InvestigationTemplate.user_id == user_id,
            )
            .first()
        )

        if not template:
            raise HTTPException(
                status_code=404, detail=f"Template {template_id} not found"
            )

        if name:
            template.name = name
        if description:
            template.description = description
        if template_json:
            InvestigationSettings(**template_json)
            template.template_json = json.dumps(template_json)
            template.version += 1
        if tags is not None:
            template.tags = json.dumps(tags)

        template.updated_at = datetime.utcnow()
        self.db.commit()
        self.db.refresh(template)

        return template.to_dict()

    def delete_template(self, template_id: str, user_id: str) -> None:
        """Delete template."""
        template = (
            self.db.query(InvestigationTemplate)
            .filter(
                InvestigationTemplate.template_id == template_id,
                InvestigationTemplate.user_id == user_id,
            )
            .first()
        )

        if not template:
            raise HTTPException(
                status_code=404, detail=f"Template {template_id} not found"
            )

        self.db.delete(template)
        self.db.commit()

    def apply_template(
        self,
        template_id: str,
        user_id: str,
        investigation_id: str,
        entity_values: Dict[str, str],
        overrides: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Apply template, replace placeholders, create investigation."""
        template = (
            self.db.query(InvestigationTemplate)
            .filter(InvestigationTemplate.template_id == template_id)
            .first()
        )

        if not template:
            raise HTTPException(
                status_code=404, detail=f"Template {template_id} not found"
            )

        settings_str = replace_placeholders(template.template_json, entity_values)
        settings_json = json.loads(settings_str)

        if overrides:
            settings_json = apply_overrides(settings_json, overrides)

        InvestigationSettings(**settings_json)

        state = InvestigationState(
            investigation_id=investigation_id,
            user_id=user_id,
            lifecycle_stage="CREATED",
            settings_json=json.dumps(settings_json),
            progress_json=None,
            status="PENDING",
            version=1,
        )

        template.usage_count += 1
        template.last_used = datetime.utcnow()

        self.db.add(state)
        self.db.commit()
        self.db.refresh(state)

        return state.to_dict()

File: olorin-front/src/shared/services/templateService.ts
/**
 * TemplateService - Template Management Client.
 * Feature: 005-polling-and-persistence
 *
 * SYSTEM MANDATE: Configuration-driven, no hardcoded values, complete error handling.
 */
import axios, { AxiosInstance } from 'axios';
import { getConfig } from '../config/env.config';
import { WizardState, InvestigationSettings } from '../types/wizardState';

// Load validated configuration
const envConfig = getConfig();
const config = {
  apiBaseUrl: envConfig.api.baseUrl,
  requestTimeoutMs: envConfig.api.requestTimeoutMs,
  // Note: AUTH_TOKEN not in schema - using process.env directly with validation
  authToken: process.env.REACT_APP_AUTH_TOKEN || ''
};

// Validate required configuration at module load
if (!config.authToken) {
  console.error('REACT_APP_AUTH_TOKEN not configured - template service authentication will fail');
}

interface Template {
  template_id: string;
  user_id: string;
  template_name: string;
  description: string | null;
  template_json: InvestigationSettings;
  tags: string[];
  usage_count: number;
  last_used: string | null;
  created_at: string;
  updated_at: string;
}

interface TemplateListResponse {
  templates: Template[];
  total: number;
  limit: number;
  offset: number;
}

interface ApplyTemplateRequest {
  entity_values: Array<{ entity_type: string; entity_value: string }>;
  overrides?: Partial<InvestigationSettings>;
}

class TemplateService {
  private client: AxiosInstance;

  constructor() {
    this.client = axios.create({
      baseURL: `${config.apiBaseUrl}/api/v1/templates`,
      timeout: config.requestTimeoutMs,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${config.authToken}`
      }
    });
  }

  async listTemplates(tags?: string[], limit: number = 20, offset: number = 0): Promise<TemplateListResponse> {
    try {
      const params: Record<string, any> = { limit, offset };
      if (tags && tags.length > 0) params.tags = tags.join(',');

      const response = await this.client.get<TemplateListResponse>('/', { params });
      return response.data;
    } catch (error) {
      this.handleError(error, 'listTemplates');
      throw error;
    }
  }

  async createTemplate(name: string, description: string | null, templateJson: InvestigationSettings, tags: string[] = []): Promise<Template> {
    try {
      const response = await this.client.post<Template>('/', {
        template_name: name,
        description,
        template_json: templateJson,
        tags
      });
      return response.data;
    } catch (error) {
      this.handleError(error, 'createTemplate');
      throw error;
    }
  }

  async getTemplate(templateId: string): Promise<Template | null> {
    try {
      const response = await this.client.get<Template>(`/${templateId}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) return null;
      this.handleError(error, 'getTemplate');
      throw error;
    }
  }

  async updateTemplate(templateId: string, updates: Partial<Omit<Template, 'template_id' | 'user_id' | 'usage_count' | 'created_at' | 'updated_at'>>): Promise<Template> {
    try {
      const response = await this.client.put<Template>(`/${templateId}`, updates);
      return response.data;
    } catch (error) {
      this.handleError(error, 'updateTemplate');
      throw error;
    }
  }

  async deleteTemplate(templateId: string): Promise<void> {
    try {
      await this.client.delete(`/${templateId}`);
    } catch (error) {
      this.handleError(error, 'deleteTemplate');
      throw error;
    }
  }

  async applyTemplate(templateId: string, request: ApplyTemplateRequest): Promise<WizardState> {
    try {
      const response = await this.client.post<WizardState>(`/${templateId}/apply`, request);
      return response.data;
    } catch (error) {
      this.handleError(error, 'applyTemplate');
      throw error;
    }
  }

  private handleError(error: unknown, operation: string): void {
    if (axios.isAxiosError(error)) {
      const status = error.response?.status;
      if (status === 404) {
        console.error(`[TemplateService] ${operation}: Not found`);
      } else if (status === 422) {
        console.error(`[TemplateService] ${operation}: Validation error`);
      } else {
        console.error(`[TemplateService] ${operation}: ${error.message}`);
      }
    } else {
      console.error(`[TemplateService] ${operation}:`, error);
    }
  }
}

export const templateService = new TemplateService();
